server/challenge/challenge.model.ts

```diff
@@ -1,134 +1,143 @@
-
-    if (!participant) {
-      throw new ServerError(404, `Nu există participantul cu id=${userId}`);
-    }
-
-    participant.lastDoneTask = lastDoneTask;
-    participant.meta = meta || null;
-
-    return saveChallenge(challenge);
-  }
-
-  static async getLastDoneTaskForUser(challengeId, userId) {
-    const challenge = await ChallengeModel.getByIdWithParticipants(challengeId);
-    if (!challenge) {
-      throw new ServerError(404, `Nu există challenge-ul cu id=${challengeId}`);
-    }
-
-    const participant = challenge
-      .participants
-      .find(participant => participant.user._id.toString() === userId);
-
-    if (!participant) {
-      return {
-        meta: null,
-        lastDoneTask: null,
-      }
-    }
-
-    // Returns a String with the last done task
-    return {
-      meta: participant.meta,
-      lastDoneTask: participant.lastDoneTask,
-    }
-  }
-}
-
-function saveChallenge(challenge) {
-  return new Promise((resolve, reject) => {
-    challenge.save((err, data) => {
-      if (err) {
-        return reject(err);
-      }
-
-      resolve(data);
-    });
-  });
-}
-
-module.exports = ChallengeModel;
\ No newline at end of file
+/* eslint-disable camelcase */
+import { Schema, model, models } from "mongoose";
+import uniqueValidator from "mongoose-unique-validator";
+import { ServerError, validateAgainstSchemaProps } from "../ServerUtils";
+import {
+  ChallengeInterface,
+  ChallengeParticipantInterface,
+} from "../types/type";
+const ChallengeParticipantSchema = new Schema<ChallengeParticipantInterface>(
+  {
+    user: { type: Schema.Types.ObjectId, ref: "User" },
+    // Can be null -> if he just started
+    lastDoneTask: { type: String, required: true },
+    // Each task may have it's own special Metadata
+    // Not used for now but something to keep in mind
+    meta: {
+      type: Object,
+    },
+  },
+  {
+    timestamps: {
+      createdAt: "submittedAt",
+      updatedAt: "updatedAt",
+    },
+  }
+);
+
+const ChallengeSchema = new Schema<ChallengeInterface>({
+  id: { type: String, required: true, unique: true },
+
+  // Have a list of task ids that server 2 purposes
+  //    > by using "lastDoneTask" we know how many has the user completed
+  //    > if later we decide to add/remove tasks, we'll know which users we need to notify and how
+  tasks: [{ type: String, required: true, unique: true }],
+  participants: [{ type: ChallengeParticipantSchema, required: true }],
+});
+
+const Challenge = models.Challenge || model("Challenge", ChallengeSchema);
+const ChallengeParticipant =
+  models.ChallengeParticipant ||
+  model("ChallengeParticipant", ChallengeParticipantSchema);
+
+class ChallengeModel {
+  static getById(id) {
+    return Challenge.findOne({ id });
+  }
+
+  static getByIdWithParticipants(id) {
+    return Challenge.findOne({ id }).populate({
+      path: "participants",
+      populate: {
+        path: "user",
```

Comment:
Single quote. Or is it because of the new configurations that this was converted to "double quotes"?